pkg/gardenlet/controller/bastion/add.go

```diff
@@ -0,0 +1,57 @@
+// Copyright (c) 2022 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bastion
+
+import (
+	operationsv1alpha1 "github.com/gardener/gardener/pkg/apis/operations/v1alpha1"
+	"github.com/gardener/gardener/pkg/controllerutils"
+	confighelper "github.com/gardener/gardener/pkg/gardenlet/apis/config/helper"
+
+	"k8s.io/client-go/util/workqueue"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/cluster"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+)
+
+// ControllerName is the name of this controller.
+const ControllerName = "bastion-controller"
+
+// AddToManager adds Reconciler to the given manager.
+func (r *Reconciler) AddToManager(mgr manager.Manager, gardenCluster, seedCluster cluster.Cluster) error {
+	if r.GardenClient == nil {
+		r.GardenClient = gardenCluster.GetClient()
+	}
+	if r.SeedClient == nil {
+		r.SeedClient = seedCluster.GetClient()
+	}
+
+	return builder.
+		ControllerManagedBy(mgr).
+		For(&operationsv1alpha1.Bastion{}, builder.WithPredicates(r.BastionPredicate(), predicate.GenerationChangedPredicate{})).
+		WithOptions(controller.Options{
+			MaxConcurrentReconciles: *r.Config.Controllers.Bastion.ConcurrentSyncs,
+			RecoverPanic:            true,
+			RateLimiter:             workqueue.DefaultControllerRateLimiter(),
+		}).
+		Complete(r)
+}
+
+// BastionPredicate returns the predicates for the Bastion.
+// It filters Bastion objects having Seed name as the SeedConfig's seed name in GardenletConfiguration.
+func (r *Reconciler) BastionPredicate() predicate.Predicate {
+	return predicate.NewPredicateFuncs(controllerutils.BastionFilterFunc(confighelper.SeedNameFromSeedConfig(r.Config.SeedConfig)))
+}
```

Comment:
Instead of this predicate, can't we centrally restrict gardenlet's cache with a field selector like this: https://github.com/gardener/gardener/blob/8f48339c5b5be7ca96072dbe680b9e7935ebe811/cmd/gardenlet/app/app.go#L278-L280